### PR TITLE
Implement ClusterRole, RoleBinding, and ClusterRoleBinding syncing

### LINKED ISSUE #63: RBAC Roles and Bindings Sync Controller
#### How would you like the feature to work?

* There are cloud `roles` and `rolebindings` that will need to be synced down to clusters. 
* A containership `role` has a list of `rules`, where each rule can be of type `cloud` or `kubernetes`. 
  * Will need to filter out only `kubernetes` related rules
  * Create a kubernetes `clusterrole` for every `role` with all of its corresponding rules
  * Use some `metadata.label` to keep track of synced `roles`
* The `rolebindings` may contain multiple `cluster_ids` that it pertains to
  * Sync will need to filter only bindings pertaining to current cluster
  * a `clusterolebinding` or `rolebinding` will need to be created on the cluster depending on if a `namespace` scope is present

#### Additional Context
* Need to fill in full details of schema @mikestoltz . This is a best effort regurgitation of our high level discussion from last week

### PR DESCRIPTION
**Pull request by mattkelly**
#### What does this pull request accomplish?

Implement ClusterRole, RoleBinding, and ClusterRoleBinding syncing based on the following logic previously discussed:

1. Sync all roles (which include rules) from sideload route to Kubernetes ClusterRoles (`name` == cloud role ID)
2. Sync all role bindings for each role to ClusterRoleBindings or RoleBindings  depending on if `namespace` is set

- TeamBinding corresponds to Kubernetes `kind = Group`, `subject` is `team_id`
- UserBinding corresponds to Kubernetes `kind = User`, `subject` is `user_id`

 #### What issue(s) does this fix?
* Closes #63

 #### Additional Considerations

 ### Testing

In progress - currently noticing a few weird things which I've inquired about in Slack

 ### Dependencies
 None

### CODE REVIEW

File: pkg/coordinator/authorization_role_binding_controller.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+		}
+
+		return errors.Wrapf(err, "getting AuthorizationRoleBinding %s for reconciliation", name)
+	}
+
+	id := authorizationRoleBinding.Spec.ID
+
+	roleNamespace := authorizationRoleBinding.Namespace
+	if roleNamespace == "" {
+		// No namespace, so this should be synced to a ClusterRoleBinding
```
Comment (mattkelly) [inline pkg/coordinator/authorization_role_binding_controller.go]: These blocks could probably be pulled out into helpers

File: pkg/coordinator/authorization_role_binding_controller.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+	return err
+}
+
+func clusterRoleBindingFromAuthorizationRoleBinding(authRoleBinding csauthv3.AuthorizationRoleBinding) rbacv1.ClusterRoleBinding {
+	return rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   authRoleBinding.Spec.ID,
+			Labels: constants.BuildContainershipLabelMap(nil),
+			OwnerReferences: []metav1.OwnerReference{
```
Comment (mattkelly) [inline pkg/coordinator/authorization_role_binding_controller.go]: Consider adding a helper to create this

File: pkg/coordinator/authorization_role_binding_controller.go [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+const (
+	authorizationRoleBindingControllerName = "AuthorizationRoleBindingController"
+
+	clusterRoleBindingDelayBetweenRetries = 30 * time.Second
+
+	maxAuthorizationRoleBindingControllerRetries = 10
+)
+
+// AuthorizationRoleBindingController is the controller implementation for the containership
+// upgrading clusters to a users desired kubernetes version
```
Comment (nicktate) [inline pkg/coordinator/authorization_role_binding_controller.go]: should this be more about rbac than cluster upgrades?

File: pkg/coordinator/authorization_role_controller.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+const (
+	authorizationRoleControllerName = "AuthorizationRoleController"
+
+	clusterRoleDelayBetweenRetries = 30 * time.Second
+
+	maxAuthorizationRoleControllerRetries = 10
+)
+
+// AuthorizationRoleController is the controller implementation for the containership
+// upgrading clusters to a users desired kubernetes version
```
Comment (nicktate) [inline pkg/coordinator/authorization_role_controller.go]: lets update this comment to refer to syncing roles / rules

File: pkg/coordinator/authorization_role_controller.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+func clusterRoleFromAuthorizationRole(authRole csauthv3.AuthorizationRole) rbacv1.ClusterRole {
+	rules := make([]rbacv1.PolicyRule, len(authRole.Spec.Rules))
+	for i, authRule := range authRole.Spec.Rules {
+		rules[i] = rbacv1.PolicyRule{
+			Verbs:         authRule.Verbs,
+			APIGroups:     authRule.APIGroups,
+			Resources:     authRule.Resources,
+			ResourceNames: authRule.ResourceNames,
```
Comment (nicktate) [inline pkg/coordinator/authorization_role_controller.go]: It's not yet part of the actual cloud auth rule, but we are going to add the following key to the `authorizationRule` 

`nonResourceURLs` : https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#policyrule-v1beta1-rbac-authorization-k8s-io

We should add it in now so when it is available, it is already good to go

File: pkg/coordinator/authorization_role_binding_controller.go (hunk 3 above)
Comment (mattkelly) [reply pkg/coordinator/authorization_role_binding_controller.go]: Probably

File: pkg/coordinator/authorization_role_controller.go (hunk 5 above)
Comment (nicktate) [reply pkg/coordinator/authorization_role_controller.go]: It's going to be set as `non_resource_urls` on the cloud api object

File: pkg/coordinator/authorization_role_binding_controller.go (hunk 1 above)
Comment (mattkelly) [reply pkg/coordinator/authorization_role_binding_controller.go]: Fiddled with this - it's not really much cleaner to pull these out, so I'm leaving as-is for now.

**Comment by mattkelly** [comment]
Added non-resource URL support, fixed a bug where I was referencing the wrong namespace on binding creation, and addressed comments. Should be good for another look.

**Comment by mattkelly** [comment]
Re-tested again this morning with what looks to be a cleaned up stage environment. Looks good to me.

**Pull request merged**